### DISCUSSION TITLE
Labs settings lost after power cycle

### DISCUSSION THREAD

**Discussion by TripleFJeff** [gopro/labs#371] [Q&A]
I recently set up a Hero 8 to use as a motorsports "dashcam" that was powered by the car and was completely automated. After many cycles of changing and testing the setup it worked great for several days at a race. Part of the setup is that it wakes on power and is set to never auto off, so once the car is shut off it eventually kills the battery. That wasn't an issue as it would still wake on power when the car was powered up again.

I had put a set of QR images into a spreadsheet so I could quickly reset and setup a camera with all of the settings, and had used it many times with no issues. I had also shared these QRs with another person who successfully setup another Hero 8 to do the same thing.

After sitting for a few weeks I plugged the camera into power, and it immediately wanted me to set the time either manually or by the app. At first I chose app but didn't actually set the time since it is supposed to set via GPS eventually. Then I went through and tried to setup the camera exactly the same as before. All of the QR codes "worked" (my burn in was showing, it enabled the start on GPS speed), but after shutting the camera off all of the settings were lost. No burn-in, no settings changes, no wake on power, and no boot command. Tried several times, manually set the date, formatted the card, I can't get it to work.

Any ideas on what the issue could be? Also, what would cause it to lose date and time or prompt me to set it? I've killed the battery repeatedly and it never prompted me to set the time.

**Comment by TripleFJeff** [comment]
For reference, this is my sequence to setup the camera:

!RESET (reset)
g1!MBRNO=10!MBURN="(0,40)[BCHH:MM:SS AA mm-dd-yyyy ]" (burn-in date and time)
mVr1080p60fSoSM25,50t0e1hS0aSg1q0dVoV0B4D4R0CS1 (basic settings)
!MWAKE=2 (wake on power)
!MBOOT="!Lspd"!SAVEspd=!SK32H30!R (boot command to start on speed)

**Reply by dnewman-gpsw** [reply]
If the battery is out or dead for more than 24 hours, that is when you can lose the clock.  Add a fresh battery, then set the clock, App, Screen or QR Code..

Try the !RESET again, and a regular factory reset, to make sure.  Failure to save is potential a memory corruption of the Labs metadata. Capture any file after a boot, and use https://gopro.github.io/labs/control/metadata/ to see if you settings are there.

Unfortunately HERO8 doesn't use GPS time to reset the clock.

**Reply by TripleFJeff** [reply]
Is there a secret decoder ring for the metadata? Not sure how to take what the metadata decoder shows and compare it to my settings.

I did a factory reset and reloaded firmware and the problem persists. 

The issue seems to be with something in the basic settings, if I do everything but that it all seems to work as expected. But once I scan the basic settings it breaks it. I know that some of the settings take-- for instance I can see my spot metering location pop up on the screen. But power off set to never doesn't work. If I scan a QR for just the power off it works alone, but not in my jumbo settings string.

I'm super confused because I am using saved screen shots of the QRs so it isn't like they can change.

**Reply by TripleFJeff** [reply]
I think I have stumbled on part of the problem. I went through and tried to find and manually setup the camera to match my "Basic Settings" string. One of the things I had to change was that Quick Capture was on (by default, since I did a reset), and once I turned it off it seemed to work, I could scan the basic settings and it wouldn't break all of it. I am still having an issue where the QR won't change the auto off. Maybe that never worked and I didn't notice it because I had already changed it manually? A simple QR for oC works, but my big one does not.

**Reply by dnewman-gpsw** [reply]
use oC0 <- the more reliable Power Off Never.

**Reply by TripleFJeff** [reply]
When you say "more reliable", you mean "actually works"? :)

I don't know how I came up with that string that just had the "C" in it rather than C0; I thought I had copied it right from the QR creator but that doesn't seem possible as "C" isn't a valid code for anything. But if I change that to C0 (and I confirmed the entire string is good by recreating it in the creator) it still didn't change the auto power off to never. If I use a QR for just "oC0" it worked, but it didn't work in my long string.

So then I went back and double checked some of the other settings and it turns out that none of the "camera preference" ones work in my string: 
Beep Off
Display Brightness 40%
LED all on
Rotation Unlocked
Auto Off Never
Screen Off 1 minute

I didn't notice it before because I already had the camera set this way; the purpose of this string was just to make sure the settings were right in case they got changed in the future.

They do work all together as "oV0B4D4R0C0S1". They just don't work as part of "mVr1080p60fSoSM25,50t0e1hS0aSg1q0dVoV0B4D4R0C0S1". I could split that into 2 different QRs as a workaround, but it isn't good if the QR creator makes ones that don't work. Any idea why it doesn't? I do have some settings in there that aren't supposed to work on a HERO 8, could that be causing it?

I think finding this confirms that having Quick Capture is what was interfering with the saving of the settings. If the camera sits long enough that it loses the clock, does that mean it will also lose all of its other settings? I had Quick Capture off, would that have been lost with the long battery off?

**Reply by dnewman-gpsw** [reply]
It appears to be a HERO8 bug.  The long string works fine on HERO10 & 11.  The commands have changed a little over the 3 years of Labs, oC become oC0 was one of the changes.  Just put oC0 first, that works.

**Reply by dnewman-gpsw** [reply]
Only the clock need some power to run, all other settings are not lost.

**Reply by TripleFJeff** [reply]
OK, I still can't account for everything that happened but I seem to have it working correctly again. Can you explain a bit about how the settings and Lab features (like overlay and wake with power) get written to non-volatile memory? Does that all happen at shutdown? Seems like maybe I should do the Labs stuff first, then shut it down, and do the settings after a restart.

**Reply by dnewman-gpsw** [reply]
!Mwxyz commands are written to Non-volatile at scan time, there is not need to shutdown to get that storage.

**Reply by TripleFJeff** [reply]
If the !M commands are written to memory on scan, how is it that the other setting scans can cause them to not stick?

**Comment by dnewman-gpsw** [comment]
Like what?

**Comment by TripleFJeff** [comment]
I just now took the camera that was setup with the Burn In, Wake on Power, and Boot Script working as it should. Ran through the sequence below; didn't go exactly as I thought it would, but managed to break it and fix it. No idea what broke it and what fixed it. Can you make any sense of this?

A few weeks ago when I first setup the camera I probably scanned hundreds of QRs while tinkering and never had a problem with things not working after reboots. I used Sequence "A" below dozens of times, and sent it to someone else who successfully used it on a HERO8. Now I can't get it to work even right after a factory reset and Firmware install.

Not to complicate things further, but another problem I've encountered when things aren't working right is that when I try to wake it on USB power, it locks it up. It doesn't wake, and then is completely unresponsive to buttons. Only way to get it turned on again is to unplug USB and take out the battery.

I ran the following sequence of QRs, call this Sequence "A":
!RESET (reset)
g1!MBRNO=10!MBURN="(0,40)[BCHH:MM:SS AA mm-dd-yyyy ]" (burn-in date and time)
mVr1080p60fSoSM25,50t0e1hS0aSg1q0dVoV0B4D4R0CS1 (basic settings, old version with incorrect "C" instead of "C0")
!MWAKE=2 (wake on power)
!MBOOT="!Lspd"!SAVEspd=!SK32H30!R (boot command to start on speed)

All of the QRs scanned successfully with the check mark, and the Burn In was showing on the screen. The boot command came up on the screen, but it didn't run (no start on speed active). I restarted the camera, and the Burn In, Wake on Power, and Boot Script now don't work.

So then I tried running it without the basic settings, call this Sequence "B":

!RESET (reset)
g1!MBRNO=10!MBURN="(0,40)[BCHH:MM:SS AA mm-dd-yyyy ]" (burn-in date and time)
!MWAKE=2 (wake on power)
!MBOOT="!Lspd"!SAVEspd=!SK32H30!R (boot command to start on speed)

When I do this the boot command actually runs, but when I restart it nothing works.

Run a factory reset and turn Quick Capture off.
Run Sequence "A", Burn In shows and Boot Script runs (start on speed active)
Restart the camera, none of it works.

Run a factory reset and turn Quick Capture off.
Run Sequence "B", Burn In shows and Boot Script runs (start on speed active)
Restart the camera, none of it works.

Run a factory reset and turn Quick Capture off.
Reinstall Firmware
Run Sequence "A", Burn In shows and Boot Script runs (start on speed active)
Restart the camera, none of it works.

Run a factory reset and turn Quick Capture off.
Reinstall Firmware
Run Sequence "B", Burn In shows and Boot Script runs (start on speed active)
Restart the camera, none of it works.

Run a factory reset and turn Quick Capture off.
!RESET (reset)
g1!MBRNO=10!MBURN="(0,40)[BCHH:MM:SS AA mm-dd-yyyy ]" (burn-in date and time)
Restart, no burn in.

!MBOOT="!Lspd"!SAVEspd=!SK32H30!R (boot command to start on speed)
Restart, boot script works!
Cancel out of start on speed
g1!MBRNO=10!MBURN="(0,40)[BCHH:MM:SS AA mm-dd-yyyy ]" (burn-in date and time)
Restart, boot script and burn in work!
Cancel out of start on speed
!MWAKE=2 (wake on power)
Power off
Plug in USB, wakes on power, boot script and burn in work!
mVr1080p60fSoSM25,50t0e1hS0aSg1q0dVoV0B4D4R0CS1 (basic settings, old version with incorrect "C" instead of "C0")
Disconnect USB Power, Power off.
Plug in USB, wakes on power, boot script and burn in work!

I'm baffled.

**Comment by dnewman-gpsw** [comment]
These are hacks. So reliability is not a priority, particularly when combined, as the permutations are crazy. Burn-ins are the most likely to cause issues, either not use or make them smaller, as they can cause crashes (60fps is harder then 30 etc.) Basically decide what is important, use only that.

**Reply by TripleFJeff** [reply]
I get that I can't expect everything to work, especially when combined. The part I'm having trouble understanding is that I have a combination of features that I know work together and have done many times, but sometimes when I try the exact same process with the exact same QRs in the exact same order it doesn't work. I'm just trying to get a handle on it so I know how to fix it when it happens, because doing the !RESET isn't enough. There definitely seems to be some interaction between the settings and the !M commands (that super long settings string that doesn't fully work in particular, I've stopped using that), and there also seems to be some corruption of the firmware, because I've gotten into situations that required the firmware to be reloaded. I've found that the only sure-fire way to get it working is:

!RESET (because the !M commands survive a firmware reload)
Factory Reset
Reload Firmware
Scan my 3 !M QRs
Restart
After that I can scan settings.

**Reply by dnewman-gpsw** [reply]
Chasing ghosts it seems.  The firmware should never need a reinstall, that is not a known failure point.  You have not demonstrated that !M settings are not taking.  Please show in the error with metadata.

**Reply by TripleFJeff** [reply]
Is there a reference that I can use to decode the metadata? I’ve looked at it but can’t decipher it.

I don’t know if it’s a matter of the settings not saving, or not running right at startup.

**Reply by TripleFJeff** [reply]
Here you go. Started with everything working correctly and made a recording:

DEVC 
  DVID USRM
  DVNM User Metadata
  STRM 
    BRNO 10
    BURN (0,40)[BCHH:MM:SS AA mm-dd-yyyy ]
    WAKE 2
    BOOT !Lspd
    USRD 1

Did a !RESET and made a recording:
DEVC 
  DVID USRM
  DVNM User Metadata

Scanned the following and made a recording:
g1!MBRNO=10!MBURN="(0,40)[BCHH:MM:SS AA mm-dd-yyyy ]"
mVr1080p60fSoSM25,50t0e1hS0aSg1q0dVoV0B4D4R0CS1
!MWAKE=2
!MBOOT="!Lspd"!SAVEspd=!SK32H30!R

DEVC 
  DVID USRM
  DVNM User Metadata
  USRD 1

Shutoff and restarted. No burn in or boot script. Made a recording.

DEVC 
  DVID USRM
  DVNM User Metadata
  USRD 1

Scanned the following and made a recording:
!MBOOT="!Lspd"!SAVEspd=!SK32H30!R

DEVC 
  DVID USRM
  DVNM User Metadata
  STRM 
    EMPT 2
    BOOT !Lspd
    USRD 

Restart and boot script works. Scanned the following and made a recording:
!MWAKE=2
g1!MBRNO=10!MBURN="(0,40)[BCHH:MM:SS AA mm-dd-yyyy ]"

DEVC 
  DVID USRM
  DVNM User Metadata
  STRM 
    BOOT !Lspd
    WAKE 2
    BRNO 10
    BURN (0,40)[BCHH:MM:SS AA mm-dd-yyyy ]
    USRD 1

Shut down and plug in USB power. Wakes on power, burn in present, boot script runs. Working as it should.

Seems like something in this sequence prevents any of it it from being saved?
g1!MBRNO=10!MBURN="(0,40)[BCHH:MM:SS AA mm-dd-yyyy ]"
mVr1080p60fSoSM25,50t0e1hS0aSg1q0dVoV0B4D4R0CS1
!MWAKE=2
!MBOOT="!Lspd"!SAVEspd=!SK32H30!R

**Reply by TripleFJeff** [reply]
@dnewman-gpsw any thoughts on this? Seem clear that there is an instance where the QRs didn't take.

**Reply by dnewman-gpsw** [reply]
Not really.  All the !M commands are working as expected.  Seems you shouldn't combine them with camera settings.  You will have to do a lot of testing, to determine which one.
